### ISSUE TITLE
getOptionValueSource() in a command context returns `undefined` for global options

### ISSUE THREAD

**Issue by sinedied** [tj/commander.js#1713]
Hello,

I need to distinguish in my code which options comes from the user (via CLI options), and which one are just the default values.
For that, I use `getOptionValueSource()` but I have an issue with that, as it always return undefined when trying to get the source of a global option in a command context.

I retrieve all options using `command.optsWithGlobals()`, but then `command.getOptionValueSource()` only works for command-level options, and if I use `program.getOptionValueSource()` it only works with global-level options.

Should I merge all options manually instead of using `optsWithGlobals()` to separate the two contexts when trying to determine the source of the options? It would seems a bit complicated for a simple use case.

Would you consider adding a `getOptionValueSourceWithGlobals` to match the behavior of `optsWithGlobals`, so that we could retrieve the option's source if it comes either from command or global level?

**Comment by shadowspawn** [comment]
> Should I merge all options manually instead of using optsWithGlobals() to separate the two contexts when trying to determine the source of the options? It would seems a bit complicated for a simple use case.

I am not sure of best approach for your case. You could possibly write your own `getOptionValueSourceWithGlobals` which searches up the command chain looking for a hit. (I haven't tried this and not sure if there are complications.)

If you have global options and local options, and whatever you are doing with them you need to tell where they came from using `.getOptionValueSource()`, I don't think you get to call it simple. 😄 

> I need to distinguish in my code which options comes from the user (via CLI options), and which one are just the default values.

For interest, can you give an insight into why you need to distinguish?

> Would you consider adding a getOptionValueSourceWithGlobals to match the behavior of optsWithGlobals, so that we could retrieve the option's source if it comes either from command or global level?

My first impression is it is not a common enough case to build into Commander. But happy to leave issue open and see if other people have similar need.

**Comment by shadowspawn** [comment]
I'll add that `.optsWithGlobals()` was added in 9.0 and `.getOptionValueSource()` in 8.3, so both fairly new and may take a while to see how they are being used and could be improved. Interesting to hear you are using both. 👍

**Comment by sinedied** [comment]
> For interest, can you give an insight into why you need to distinguish?

I'm working on a CLI that can use options from a config file, that I load when the CLI is invoked. But I need to be able to override any options from the config file using CLI options, hence this need. I would have thought that the use case is pretty common 😄 

At the command level, I need both global and command options to execute the command, and I process all options by merging the options that comes from the config file, then from the CLI. As a lot of options have default values, I need to only keep CLI options that were explicitly set by the user to override the ones from the config file.

BTW, I found a simple workaround to implement `getOptionValueSourceWithGlobals()`:
```
const getOptionValueSourceWithGlobals(option, command) => command. getOptionValueSource(option) ?? program.getOptionValueSource(option);
```
This works as `getOptionValueSource` returns `undefined` when the option is not found in the current context.

**Comment by michaelAtCoalesce** [comment]
im having this exact same issue. 

@sinedied can you give a bit more context on what option and command are in your little code snippet? is option a string?

**Comment by shadowspawn** [comment]
Yes, in that snippet `option` is a string and `command` is the Command object as passed to the action handler (or saved when the command created).

**Comment by shadowspawn** [comment]
Opened a PR to implement `.getOptionValueSourceWithGlobals()`.

**Comment by shadowspawn** [comment]
Added in Commander v9.5.0
